`General`: Limit scope of active students line charts by course start and end date

willergeorg opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server..
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We currently provide two line charts displaying the amount of active students per week. Both do not take the course start and end date into account. This has the effect, that for example weeks before the official start of the course are displayed or the active students are shown up to the current week even though the course ended weeks ago. This is changed by this PR. 
### Description
<!-- Describe your changes in detail -->
I changed the chart implementation so they are not rendered if the start date is not passed yet and only contain data until the course end date if it is passed.
Therefore I also adapted the server implementation. In case that the end date of a course is already passed, the server only sends the data collected until the end date to the client.
The data sent is also adapted if the visible time span differs from the normal one, e.g. the start date was 1 week ago or the end course life time is shorter than the chart time span.
### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 1 Course with an active students distribution

1. Log in to Artemis
2. Navigate to Course Administration
3. Inspect the active students chart for the course. Make sure the values are displayed appropriately. Do the same for the active students chart on course detail page.
4. Edit the course and set the end date e.g. 2 weeks before the current date.
5. Check both charts again. The data should switch accordingly and the data labels should reflect the fact, that old data is displayed (x weeks ago as most right data label on course management and the corresponding ISO week on course detail).
6. Edit the course again. Delete the end date and set the start date on some point in future
7. Check the charts. No charts should be displayed, instead a lettering explains that the course did not start yet.
8. Edit the course again. Set the start date on some point later than 3 weeks before the current date.
9. Check both charts. Now the charts should start from the week of the dedicated start date.
10. Edit the course again. Set the start date more than 16 weeks before the current date. Check the charts. The chart on course management should not be affected by it. The chart on course detail page should only allow going back in time until the start date (by clicking the left arrow)
11. Edit the course again. Set the start date of the course e.g. 2 weeks ago and the end date of the course e.g. 1 week ago. Return to the charts and make sure they change appropriately.
12. Go wild with different combinations of start and end date and make sure no unexpected behavior occurs.

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->
| File name | Branch coverage (%) | Line coverage (%) |
| -- | -- | -- |
| `CourseRescource.java` | 73 | 93 |
| `CourseService.java` | 75 | 90 |
| `course-detail-line-chart.component.ts` | 79.16 | 94.73 |
| `course-management-overview-statistics.component.ts` | 91.66 | 91.48 |


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
### Example screenshot how the course management chart should look after step 4:
![image](https://user-images.githubusercontent.com/80622272/157659914-944df84b-d39e-4d09-9cef-036824296911.png)
For course detail (current date is in CW 10):
![image](https://user-images.githubusercontent.com/80622272/157843401-1702602a-9bb9-404d-aa32-26f8ec0ae2de.png)

### Example after step 6 (similar for course detail):
![image](https://user-images.githubusercontent.com/80622272/164494675-ef70a937-f3bf-4ebb-bb2b-a4ec4660a6f9.png)

### Example after step 8:
![image](https://user-images.githubusercontent.com/80622272/157845724-bf7d60b1-b9a5-4e8a-82c7-2f4a81955450.png)
For course detail:
![image](https://user-images.githubusercontent.com/80622272/157843835-4175d64b-5a6c-49fb-bb3a-8c84d6d05106.png)

<details>
     <summary>Example for course detail after step 10:</summary>

https://user-images.githubusercontent.com/80622272/157844612-a29f249f-b1c3-43f1-bb44-21117589f43c.mp4


</details>

### Example for course management after step 11:
![image](https://user-images.githubusercontent.com/80622272/157844946-b1667270-4db9-495a-a674-62d77d8aa351.png)
For course detail:
![image](https://user-images.githubusercontent.com/80622272/157845013-03972ec4-b7c0-4451-a228-eaa48a2a8eca.png)

canberkanar commented:
For testing step 4;
I saw that the chart on course management behaved accordingly. However, the chart within course details is stalled.
I did the testing on my local setup, with a course that previously existed. First I uploaded a file as a student to trigger an activity and confirmed it on the tables. Then I set the course end date two weeks prior to today.
![image](https://user-images.githubusercontent.com/67344158/158162099-f472997c-3ae6-4930-902f-a2753897e501.png)
![image](https://user-images.githubusercontent.com/67344158/158162154-6b95b0d3-9f2b-4842-b1be-5312da150808.png)

canberkanar commented:
Another problematic setup;
![image](https://user-images.githubusercontent.com/67344158/158163716-a3dee556-7708-421d-9723-602e3d195036.png)
![image](https://user-images.githubusercontent.com/67344158/158163790-14584597-82fb-4375-8adf-c065d762b8c9.png)
![image](https://user-images.githubusercontent.com/67344158/158163811-37ddd93f-9a9b-4497-92b4-1e1ee3bfbd9d.png)

canberkanar reviewed (dismissed):
Tested locally and on ts1. On local setup, the chart on course details page was loading infinitely but everything was as expected on the test server.

b-fein commented on src/main/java/de/tum/in/www1/artemis/service/CourseService.java:
```diff
@@ -674,4 +675,39 @@ public void checkIfEditorGroupsNeedsToBeCreated(Course course) {
+    public int determineTimeSpanSizeForActiveStudents(Course course, ZonedDateTime endDate, int maximalSize) {
+        var spanTime = maximalSize;
+        if (course.getStartDate() != null) {
+            var startDateIsoWeek = statisticsRepository.getWeekOfDate(course.getStartDate());
+            var endDateIsoWeek = statisticsRepository.getWeekOfDate(endDate);
+            int weeksInYear = Math.toIntExact(IsoFields.WEEK_OF_WEEK_BASED_YEAR.rangeRefinedBy(course.getStartDate()).getMaximum());
+            int amountOfWeeksBetween = (endDateIsoWeek - startDateIsoWeek + weeksInYear) % weeksInYear;
+            spanTime = Math.min(maximalSize, amountOfWeeksBetween + 1);
```
I think there was something similar some time ago but I don’t remember the details: Why does
```java
course.getStartDate().until(endDate.plusWeeks(1), ChronoUnit.WEEKS)
```
not work in this case?

willergeorg replied:
I oriented at code that already performed a similar task. I tried your suggestion, which seemed to work for me, and passed the adapted test cases. Changed accordingly in b9ecc6d

b-fein commented on src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts:
```diff
@@ -121,13 +126,18 @@ export class CourseDetailLineChartComponent implements OnChanges {
         const prefix = this.translateService.instant('calendar_week');
-        const startDate = dayjs().subtract(this.displayedNumberOfWeeks - 1 + this.displayedNumberOfWeeks * -this.currentPeriod, 'weeks');
-        const endDate = this.currentPeriod !== 0 ? dayjs().subtract(this.displayedNumberOfWeeks * -this.currentPeriod, 'weeks') : dayjs();
+        const endDate = dayjs().subtract(this.currentOffsetToEndDate + this.displayedNumberOfWeeks * -this.currentPeriod, 'weeks');
```
`this.currentOffsetToEndDate + this.displayedNumberOfWeeks * -this.currentPeriod` appears in the calculation of the `startDate` again (with some other addition in between). Can you move it to a variable with a name that explains what is actually calculated here?

willergeorg replied:
Done in b9ecc6d

b-fein commented on src/main/webapp/app/shared/chart/active-students-chart.directive.ts:
```diff
@@ -0,0 +1,66 @@
+import { Course } from 'app/entities/course.model';
+import dayjs from 'dayjs/esm';
+
+@Directive()
+export class ActiveStudentsChartDirective {
```
Does this actually need to be a directive or could it be a regular class?

willergeorg replied:
Indeed, should be fixed

b-fein commented on src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java:
```diff
@@ -1620,6 +1620,7 @@ public void testGetExerciseStatsForCourseOverview() throws Exception {
         assertThat(courseDtos).hasSize(1);
         var dto = courseDtos.get(0);
         assertThat(dto.getCourseId()).isEqualTo(instructorsCourse.getId());
+        assertThat(dto.getActiveStudents().size()).as("course was only active for 3 days").isEqualTo(1);
```
```suggestion
        assertThat(dto.getActiveStudents()).as("course was only active for 3 days").hasSize(1);
```

willergeorg replied:
This seems to be better to me, thanks for the suggestion.

b-fein reviewed (approved):
Code review: Looks good, though with those date calculation is is likely that small issues like off-by-ones are only noticed during manual testing.
I have left a few comments which are mostly not suggestions but questions to help my understanding.

willergeorg commented:
> Code review: Looks good, though with those date calculation is is likely that small issues like off-by-ones are only noticed during manual testing. I have left a few comments which are mostly not suggestions but questions to help my understanding.

Thanks for taking the time and looking at the changes. I updated the code according to your feedback.

b-fein reviewed (approved):
Thanks, code looks good! :+1:

Dominik-Weinzierl reviewed (changes requested):
I noticed some issues after step 4 (course with zero participants). Instead of `NaN%` it should be `0%`.

<details>
<summary>Before step 4:</summary>

![image](https://user-images.githubusercontent.com/60005702/165156319-930dedf7-80d0-469f-8b91-f127c164d381.png)
</details>


<details>
<summary>After step 4:</summary>

![image](https://user-images.githubusercontent.com/60005702/165155773-a1dd7073-2e4a-491e-aabd-24d58f5ebad4.png)
</details>

Edit: This can also be solved in a follow up PR if you want to merge this PR :+1:

willergeorg commented:
> I noticed some issues after step 4 (course with zero participants). Instead of `NaN%` it should be `0%`.
> 
> Before step 4:
> After step 4:
> Edit: This can also be solved in a follow up PR if you want to merge this PR 👍

Thanks for reporting this issue to me. I refactored the corresponding client code so that NaN should not occur anymore.
I also changed the server test implementation a little bit, because the prior configuration of submission dates led to a flaky server test. If you detect anything else that seems to be unintended, just let me know.

bassner commented on src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts:
```diff
@@ -121,13 +127,26 @@ export class CourseDetailLineChartComponent implements OnChanges {
+        this.currentSpanSize = Math.min(remainingWeeksTillStartDate, this.displayedNumberOfWeeks);
+        // for the start date, we subtract the currently possible span size - 1 from the end date in addition
+        const startDate = dayjs().subtract(diffToLastChartWeek + this.currentSpanSize - 1, 'weeks');
+        this.startDateDisplayed = this.course.startDate ? remainingWeeksTillStartDate <= this.displayedNumberOfWeeks : false;
```
```suggestion
        this.startDateDisplayed = !!this.course.startDate && remainingWeeksTillStartDate <= this.displayedNumberOfWeeks;
```

bassner reviewed (changes requested):
Only a few minor _suggestions_ :)

Dominik-Weinzierl reviewed (approved):
Well done :tada: the problem with `NaN` is solved :+1: (tested locally)

Dominik-Weinzierl commented on src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.html:
```diff
@@ -7,57 +7,70 @@ <h5>{{ 'artemisApp.courseStatistics.activeStudents' | artemisTranslate }}</h5>
+                class="col-md-1 d-flex justify-content-start align-items-center"
+                role="button"
+                (click)="switchTimeSpan(RIGHT)"
+                [ngClass]="showsCurrentWeek ? 'hide-arrow' : ''"
```
You could also use `[class.hide-arrow]="showsCurrentWeek"` - but both ways are good :+1:

akesfeden commented:
Tested locally, I am not sure if this is related to your changes but when the ratio is 100%, it seems like the line goes over the top of the chart. I would expect the line to be continuous. Can you also reproduce this?

![image](https://user-images.githubusercontent.com/18427209/165421712-6c07c28f-e491-41c8-baca-91d007c1bf0a.png)

willergeorg commented:
> Tested locally, I am not sure if this is related to your changes but when the ratio is 100%, it seems like the line goes over the top of the chart. I would expect the line to be continuous. Can you also reproduce this?
> 
> ![image](https://user-images.githubusercontent.com/18427209/165421712-6c07c28f-e491-41c8-baca-91d007c1bf0a.png)

Thanks for reporting it. This is a known issue unrelated to this PR. The computation of the relative value is based on the current number of enrolled students. If you remove students from the course, the relative value can exceed 100%. As students currently cannot leave a course by themselves and the instructor would need to remove them manually, this is an edge case on production, especially for a large number of students.

cremertim reviewed (dismissed):
tested locally according to the testing steps. Works as expected for me

bassner reviewed (dismissed):
LGTM 👍

b-fein commented on src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts:
```diff
@@ -121,13 +127,26 @@ export class CourseDetailLineChartComponent implements OnChanges {
+        for (let i = 0; i < this.currentSpanSize; i++) {
             currentWeek = dayjs()
-                .subtract(this.displayedNumberOfWeeks - 1 + this.displayedNumberOfWeeks * -this.currentPeriod - i, 'weeks')
+                .subtract(this.currentOffsetToEndDate + this.currentSpanSize - 1 + this.displayedNumberOfWeeks * -this.currentPeriod - i, 'weeks')
```
```suggestion
                .subtract(this.currentOffsetToEndDate + this.currentSpanSize - 1 - this.displayedNumberOfWeeks * this.currentPeriod - i, 'weeks')
```

akesfeden reviewed (approved):
Thank you for the thorough explanation @willergeorg, besides that I haven't seen any issues in my local test.

b-fein reviewed (approved):
Code: Changes since last review look good.

filip-gregurevic reviewed (approved):
Re-approve 👍🏼

bassner reviewed (approved):
Changes since last review make sense 👍

Dominik-Weinzierl reviewed (approved):
Changes since last approve look good :+1:   tested again locally and could not find any new problems. Well done :tada:

akesfeden reviewed (approved):
Code changes look good to me.
